c/tskit/trees.c

```diff
@@ -2103,22 +2126,635 @@ tsk_treeseq_sample_count_stat(const tsk_treeseq_t *self, tsk_size_t num_sample_s
+            tj++;
+            u = edge_child[h];
+            v = edge_parent[h];
+            parent[u] = v;
+
+            c = right_child[v];
+            if (c == TSK_NULL) {
+                left_sib[u] = TSK_NULL;
+                right_sib[u] = TSK_NULL;
+            } else {
+                right_sib[c] = u;
+                left_sib[u] = c;
+                right_sib[u] = TSK_NULL;
+            }
+
+            right_child[v] = u;
+        }
+        t_right = sequence_length;
+        if (tj < num_edges) {
+            t_right = TSK_MIN(t_right, edge_left[edges_in[tj]]);
+        }
+        if (tk < num_edges) {
+            t_right = TSK_MIN(t_right, edge_right[edges_out[tk]]);
+        }
+
+        ret = get_mutation_samples(self, tree_index, num_sample_chunks, right_child,
+            left_sib, parent, site_offsets, allele_samples, &num_alleles);
+        if (ret != 0) {
+            goto out;
+        }
+
+        tree_index++;
+        t_left = t_right;
+    }
+out:
+    if (parent != NULL) {
+        free(parent);
+    }
+    if (right_child != NULL) {
+        free(right_child);
+    }
+    if (left_sib != NULL) {
+        free(left_sib);
+    }
+    if (right_sib != NULL) {
+        free(right_sib);
+    }
+    return ret;
+}
+
+static int
+tsk_treeseq_two_site_general_stat(const tsk_treeseq_t *self, tsk_size_t state_dim,
+    const double *sample_weights, tsk_size_t result_dim, general_stat_func_t *f,
+    sample_count_stat_params_t *f_params, norm_func_t *norm_f,
+    const double *TSK_UNUSED(left_window), const double *TSK_UNUSED(right_window),
+    tsk_flags_t options, tsk_size_t *result_size, double **result)
+{
+    int ret = 0;
+    const tsk_size_t num_sites = self->tables->sites.num_rows;
+    const tsk_size_t num_samples = self->num_samples;
+    const tsk_size_t max_alleles = self->tables->mutations.num_rows + num_sites;
+    tsk_size_t num_sample_chunks = TSK_BIT_ARRAY_NUM_CHUNKS(num_samples);
+    tsk_size_t *restrict site_offsets = tsk_malloc(num_sites * sizeof(*site_offsets));
+    tsk_size_t *restrict num_alleles = tsk_malloc(num_sites * sizeof(*num_alleles));
+    tsk_bit_array_t *restrict sample_sets
+        = tsk_calloc(num_sample_chunks * state_dim, sizeof(*sample_sets));
+    tsk_bit_array_t *allele_samples
+        = tsk_calloc(max_alleles * num_sample_chunks, sizeof(*allele_samples));
+    tsk_bit_array_t all_samples_bits[num_sample_chunks];
+
+    if (site_offsets == NULL || num_alleles == NULL || sample_sets == NULL
+        || allele_samples == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // A future improvement could get a union of all sample sets
+    // instead of all samples
+    get_all_samples_bits(all_samples_bits, num_samples, num_sample_chunks);
+
+    tsk_bit_array_t *allele_samples_row;
+    tsk_size_t num_alleles_cumsum = 0;
+    for (tsk_size_t t = 0; t < self->num_trees; t++) {
+        for (tsk_size_t s = 0; s < self->tree_sites_length[t]; s++) {
```

Comment:
This could just loop over the sites - the sites in the table are in the same order as encountered when iterating over (trees; sites within trees).

Reply:
Oh, this is good to know, I will make this change as well

Reply:
See https://tskit.dev/tskit/docs/latest/data-model.html#valid-tree-sequence-requirements (and https://tskit.dev/tskit/docs/latest/data-model.html#table-definitions ) for ordering requirements on the tables.